perfetto: add package for tracing

hijaz-siemens opened the pull request:
Perfetto package is included if tracing is enabled. Added a patch to resolve format issue for ftrace.

JIRA-ID: SB-23723

hijaz-siemens committed:
perfetto: add package for tracing

Perfetto package is included in tools-profile if tracing
is enabled. Added a patch to resolve format issue for ftrace.

JIRA-ID: SB-23723

Signed-off-by: Hasan Ijaz <hasan.ijaz@siemens.com>

hashraf91 commented on meta-sokol-flex-distro/conf/distro/sokol-flex.conf:
```diff
@@ -217,6 +217,7 @@ FEATURE_PACKAGES_tools-benchmark ?= "packagegroup-tools-benchmark"
 FEATURE_PACKAGES_multimedia      ?= "gstreamer1.0 gstreamer1.0-plugins-base gstreamer1.0-plugins-good gstreamer1.0-plugins-bad"
 FEATURE_PACKAGES_virtualization  ?= "docker-moby python3-docker-compose"
 FEATURE_PACKAGES_codebench-debug ?= "gdbserver strace openssh-sftp-server"
+FEATURE_PACKAGES_tools-profile:append:feature-tracing ?= " perfetto"
```
Drop the soft assignment and instead just set the value, as you're using :append here.

hashraf91 replied:
@Noor-Ahsan Should we move this line to distro layer from BSP layer?
```
FEATUREOVERRIDES .= "${@bb.utils.contains('DISTRO_FEATURES', 'tracing', ':feature-tracing', '', d)}"
```
As now we are supporting tracing at DISTRO level instead of a BSP feature.

Noor-Ahsan replied:
either we should use flex_os override here or create bbappend of packagegroup-core-tools-profile.bb and add perfetto in the recipe

Noor-Ahsan replied:
we already have one bbappend at ./meta-sokol-flex-distro/recipes-core/packagegroups/packagegroup-core-tools-profile.bbappend. update that bbappend

Noor-Ahsan commented on meta-sokol-flex-distro/recipes-devtools/perfetto/perfetto.bbappend:
```diff
@@ -0,0 +1,6 @@
+# ---------------------------------------------------------------------------------------------------------------------
+
+FILESEXTRAPATHS:prepend:sokol-flex := "${THISDIR}/${PN}:"
+SRC_URI += "file://0001-ftrace-Avoid-crashing-if-format-doesn-t-match-expect.patch"
```
use :prepend:sokol-flex instead of += here

hashraf91 commented:
@hijaz-siemens please rebase this PR and use 'flex-os' override in-place of 'sokol-flex'.

ahsanhussain commented on meta-flex-distro/recipes-core/packagegroups/packagegroup-core-tools-profile.bbappend:
```diff
@@ -4,3 +4,6 @@
 SYSTEMTAP:flex-os = ""
+
+# Add perfetto to tools-profile for tracing
+RDEPENDS:${PN}:append:flex-os = " perfetto"
```
Shouldn't we add a check for tracing user/distro feature because otherwise it's unconditionally changing the behavior of an upstream packagegroup.

Noor-Ahsan replied:
well, it is debatable. From other viewpoint it is saying our distro define core tools profile like this. This recipe will only be included in our image when tracing is enabled. So current solution looks fine to me.

ahsanhussain replied:
OK, let's merge it then.

hashraf91 commented on meta-flex-distro/recipes-devtools/perfetto/perfetto/0001-ftrace-Avoid-crashing-if-format-doesn-t-match-expect.patch:
```diff
@@ -0,0 +1,42 @@
+From 989cecf48f5785a35925e3c818cb968344334ac3 Mon Sep 17 00:00:00 2001
+From: Daniele Di Proietto <ddiproietto@google.com>
+Date: Tue, 9 May 2023 16:10:03 +0000
+Subject: [PATCH 1/1] ftrace: Avoid crashing if format doesn't match
```
Seems like perfetto uses FTRACE functionality from Linux kernel. Can you confirm if we need to enable this kernel configuration? Maybe by trying out some example related to function tracing and see if functions' entry and exit points are being displayed properly.

hashraf91 replied:
https://gist.github.com/johnstultz-work/0ec4974e0929c4707bfd89c876ae4735#kernel-config

hijaz-siemens committed:
perfetto: enable kernel function tracing

Kernel configs are enabled to trace functions and
function graphs in perfetto. These kernel configs
are added based on the user feature tracing.

JIRA-ID: SB-23723

Signed-off-by: Hasan Ijaz <hasan.ijaz@siemens.com>

hashraf91 commented on meta-flex-distro/recipes-kernel/linux/linux-yocto_%.bbappend:
```diff
@@ -0,0 +1,13 @@
+# Unpublished work. Copyright 2024 Siemens
+# ---------------------------------------------------------------------------------------------------------------------
+
+
```
Please drop this extra line.

hashraf91 commented on meta-flex-distro/recipes-kernel/linux/linux-yocto_%.bbappend:
```diff
@@ -0,0 +1,13 @@
+SRC_URI:append:feature-tracing = " \
+	file://ftrace.cfg \
+"
+	
```
Same here, an extra line at the end of file.

hijaz-siemens committed:
perfetto: enable sdk for userland tracing

libperfetto is added to tools-profile for userland
tracing.

JIRA-ID : SB-23723

Signed-off-by: Hasan Ijaz <hasan.ijaz@siemens.com>

hashraf91 commented on meta-flex-distro/recipes-core/packagegroups/packagegroup-core-tools-profile.bbappend:
```diff
@@ -4,3 +4,6 @@
 SYSTEMTAP:flex-os = ""
+
+# Add perfetto to tools-profile for tracing
+RDEPENDS:${PN}:append:flex-os = " perfetto"
```
Will you be adding `libperfetto`?

hashraf91 commented on meta-flex-distro/recipes-kernel/linux/linux-yocto_%.bbappend:
```diff
@@ -0,0 +1,13 @@
+
+
+FILESEXTRAPATHS:prepend := "${THISDIR}/${PN}:"
+SRC_URI:append:feature-tracing = " \
```
flex-os override should be applied here, just as you did in other recipe. If additionally you need to use tracing, you can add check with bb.utils.contains() to see if tracing is enabled in DISTRO_FEATURES.

hashraf91 commented:
Getting this warning with your changes, occurs on new build. Please see if this can be easily resolved, though it's coming from upstream recipe:
```
WARNING: perfetto-31.0-r0 do_package_qa: QA Issue: File /usr/bin/.debug/tracebox in package perfetto-dbg contains reference to TMPDIR [buildpaths]
```

hijaz-siemens committed:
perfetto: avoid warning for references to the build environment

On building the upsteam perfetto recipe, this warning is received
"WARNING:perfetto-31.0-r0 do_package_qa: QA Issue: File
/usr/bin/.debug/tracebox in package perfetto-dbg contains reference
to TMPDIR [buildpaths]". It is resolved by appending ${DEBUG_PREFIX_MAP}
to TUNE_CCARGS to sanitize debug information. The solution is
provided by this patch:
https://patchwork.yoctoproject.org/project/oe/patch/20240808112045.84285-1-alperyasinak1@gmail.com/

JIRA-ID: SB-23723

Signed-off-by: Hasan Ijaz <hasan.ijaz@siemens.com>

hashraf91 commented on meta-flex-distro/recipes-devtools/perfetto/perfetto.bbappend:
```diff
@@ -4,3 +4,6 @@
 SRC_URI:prepend:flex-os = "file://0001-ftrace-Avoid-crashing-if-format-doesn-t-match-expect.patch "
+
+# Remove build path references from debug symbols and avoid QA warnings
+TUNE_CCARGS:append = " ${DEBUG_PREFIX_MAP}" # TODO: remove when upstream is updated
```
Add the TODO comment on separate line, this gives parsing error during build.
